autopilot/contractor.go

```diff
@@ -460,18 +470,28 @@ func (c *contractor) performWalletMaintenance(ctx context.Context) error {
 	return nil
 }
 
-func (c *contractor) runContractChecks(ctx context.Context, w Worker, contracts []api.Contract, minScore float64) (toKeep []types.FileContractID, toArchive, toStopUsing map[types.FileContractID]string, toRefresh, toRenew []contractInfo, _ error) {
+func (c *contractor) runContractChecks(ctx context.Context, w Worker, contracts []api.Contract, inCurrentSet map[types.FileContractID]struct{}, minScore float64) (toKeep []types.FileContractID, toArchive, toStopUsing map[types.FileContractID]string, toRefresh, toRenew []contractInfo, _ error) {
 	if c.ap.isStopped() {
 		return
 	}
 	c.logger.Debug("running contract checks")
 
+	// convenience variables
+	state := c.ap.State()
+
+	// calculate 'maxKeepLeeway' which defines the amount of contracts we'll be
+	// lenient towards when we fail to either fetch a valid price table or the
+	// contract's revision
+	maxKeepLeeway := addLeeway(state.cfg.Contracts.Amount, 1-leewayPctRequiredContracts)
```

Comment:
It just came to me that making this relative to the size of the contract set might result in an edge case for large sets.
e.g. if `maxKeepLeeway` exceeds the number of parity pieces of an uploaded slab, we might lose that slab if all contracts turn out to be actually bad. 

I'm not sure if this is an actual issue though since the offline check should eventually get rid of them if they are not responsive. But there might be edge cases where the host can't provide a revision but the scan succeeds.

Reply:
I think this is something we need to measure and then sort by. We can easily measure "important" contracts. I think it's more or less safe to assume though that migrations would migrate it away to contract-set-contracts before it gets lost. This stuff only decides what's in the set, but we can still download from all hosts we formed a contract with.